### PR TITLE
New Joe, Pangolin Pools for week of Feb 06 2022

### PR DESCRIPTION
**Pull request by zeroecksYeti**
Added AVAX-GMX and AVAX-DOMI strats
Moved `_takeFeeRewardToSnob()` and `_swapBaseToToken()` to joe-rush-farm-base.sol

### CODE REVIEW

File: contracts/strategies/strategy-joe-rush-farm-base.sol [hunk 1]
```diff
@@ -126,4 +126,34 @@ abstract contract StrategyJoeRushFarmBase is StrategyBase {
...
+        address[] memory path = new address[](3);
+        path[0] = token1;
+        path[1] = wavax;
+        path[2] = token2;
+        IERC20(token1).safeApprove(joeRouter, 0);
+        IERC20(token1).safeApprove(joeRouter, _amount);
+        _swapTraderJoeWithPath(path, _amount);
+    }
+
+    function _takeFeeRewardToSnob(uint256 _keep, address reward) internal {
```
Comment (zeroecksYeti) [inline contracts/strategies/strategy-joe-rush-farm-base.sol:140]: Could we remove `_takeFeeJoeToSnob()` (line 113) and use this function instead for Joe rewards? Functionality is identical.

Comment (AbigailTCameron) [reply contracts/strategies/strategy-joe-rush-farm-base.sol:140]: no, because the _takeFeeJoeToSnob also has revenueShare, feeDistributor logic, etc.

Comment (fuzzylemma) [reply contracts/strategies/strategy-joe-rush-farm-base.sol:140]: this requires testing every contract that inherits from the this base

**Review by AbigailTCameron — DISMISSED** [review]
LGTM!

File: contracts/strategies/strategy-joe-rush-farm-base.sol (hunk 1 above)
Comment (Auroter) [reply contracts/strategies/strategy-joe-rush-farm-base.sol:140]: You could also simplify the body of _takeFeeJoeToSnob(_reward) to just call takeFeeRewardToSnob(joe, _reward).  Then child/other contracts can still call either method.

**Comment by jomarip** [comment]
Last comment was Looks Good to Me from @AbigailTCameron . Is there anything additional happening with this @zeroecksYeti ?

**Comment by zeroecksYeti** [comment]
> Last comment was Looks Good to Me from @AbigailTCameron . Is there anything additional happening with this @zeroecksYeti ?

Was good to go until new Joe pools dropped today, I've added them to this PR

File: contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            }
+
+            _joe = IERC20(joe).balanceOf(address(this));
+        }
+
+        // In the case of WAVAX Rewards, swap half WAVAX for USDC
+        if(_wavax > 0){
+            IERC20(wavax).safeApprove(joeRouter, 0);
+            IERC20(wavax).safeApprove(joeRouter, _wavax.div(2));   
+            _swapTraderJoe(wavax, usdc, _wavax.div(2));
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol:79]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        // In the case of USDC Rewards, swap half USDC for WAVAX
+        if(_usdc > 0){
+            IERC20(usdc).safeApprove(joeRouter, 0);
+            IERC20(usdc).safeApprove(joeRouter, _usdc.div(2));   
+            _swapTraderJoe(usdc, wavax, _usdc.div(2));
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol:86]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+          
+        }
+
+        // In the case of JOE Rewards, swap JOE for WAVAX and USDC        
+        if(_joe > 0){    
+            IERC20(joe).safeApprove(joeRouter, 0);
+            IERC20(joe).safeApprove(joeRouter, _joe);
+            _swapTraderJoe(joe, wavax, _joe.div(2));
+            _swapBaseToToken(_joe.div(2), joe, usdc);
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-avax-usdc.sol:95]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        // Take Avax Rewards    
+        uint256 _avax = address(this).balance;              // get balance of native AVAX
+        if (_avax > 0) {                                    // wrap AVAX into ERC20
+            WAVAX(wavax).deposit{value: _avax}();
+        }
+        
+        // 10% is sent to treasury
+        uint256 _wavax = IERC20(wavax).balanceOf(address(this));  // get balance of WAVAX
+        uint256 _usdce = IERC20(usdce).balanceOf(address(this));    // get balance of USDCe
+        uint256 _ust = IERC20(usdce).balanceOf(address(this));    // get balance of UST
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:46]: uint256 _ust = IERC20(**ust**).balanceOf(address(this));

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            IERC20(wavax).safeApprove(joeRouter, _wavax);   
+            _swapTraderJoe(wavax, usdce, _wavax.div(2));
+            _swapTraderJoe(wavax, ust, _wavax.div(2));
+        }
+
+        // In the case of USDCe Rewards, swap half USDCe for UST
+        if(_usdce > 0){
+            IERC20(usdce).safeApprove(joeRouter, 0);
+            IERC20(usdce).safeApprove(joeRouter, _usdce.div(2));   
+            _swapTraderJoe(usdce, ust, _usdce.div(2));
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:98]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            IERC20(usdce).safeApprove(joeRouter, 0);
+            IERC20(usdce).safeApprove(joeRouter, _usdce.div(2));   
+            _swapTraderJoe(usdce, ust, _usdce.div(2));  
+        }
+
+        // In the case of UST Rewards, swap half UST for USDCe
+        if(_ust > 0){
+            IERC20(ust).safeApprove(joeRouter, 0);
+            IERC20(ust).safeApprove(joeRouter, _ust.div(2));   
+            _swapTraderJoe(ust, usdce, _ust.div(2));
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:105]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            IERC20(ust).safeApprove(joeRouter, 0);
+            IERC20(ust).safeApprove(joeRouter, _ust.div(2));   
+            _swapTraderJoe(ust, usdce, _ust.div(2)); 
+        }
+
+        // In the case of JOE Rewards, swap JOE for USDCe and UST        
+        if(_joe > 0){    
+            IERC20(joe).safeApprove(joeRouter, 0);
+            IERC20(joe).safeApprove(joeRouter, _joe);
+            _swapTraderJoe(joe, usdce, _joe.div(2));
```
Comment (AbigailTCameron) [inline contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:112]: path might not work

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol (hunk 6 above)
Comment (zeroecksYeti) [reply contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:98]: Both tokens have avax-<token> pools; after checking the `_swapTraderJoe()` function I think it should work

```
function _swapTraderJoe(
        address _from,
        address _to,
        uint256 _amount
    ) internal {
        require(_to != address(0));

        address[] memory path;

        if (_from == wavax || _to == wavax) {
            path = new address[](2);
            path[0] = _from;
            path[1] = _to;
        } else {
            path = new address[](3);
            path[0] = _from;
            path[1] = wavax;
            path[2] = _to;
        }

        IJoeRouter(joeRouter).swapExactTokensForTokens(
            _amount,
            0,
            path,
            address(this),
            now.add(60)
        );
    }

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol (hunk 7 above)
Comment (zeroecksYeti) [reply contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:105]: see above

File: contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol (hunk 8 above)
Comment (zeroecksYeti) [reply contracts/strategies/traderJoe/strategy-joe-usdce-ust.sol:112]: see above

**Review by AbigailTCameron — APPROVED** [review]
LGTM!

**Pull request merged**